internal/crypto/cms/signed.go

```diff
@@ -0,0 +1,272 @@
+	for _, signerInfo := range d.SignerInfos {
+		if signerInfo.Version != 1 {
+			// Only IssuerAndSerialNumber is supported currently
+			return nil, VerificationError{Message: fmt.Sprintf("invalid signer info version: only version 1 is supported; got %d", signerInfo.Version)}
+		}
+
+		cert, err := d.verify(&signerInfo, &opts)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+		thumbprint, err := hashutil.ComputeHash(crypto.SHA256, cert.Raw)
+		if err != nil {
+			return nil, err
+		}
+		verifiedSignerMap[hex.EncodeToString(thumbprint)] = cert
+	}
+	if len(verifiedSignerMap) == 0 {
+		return nil, lastErr
+	}
+
+	verifiedSigningCertificates := make([]*x509.Certificate, 0, len(verifiedSignerMap))
+	for _, cert := range verifiedSignerMap {
+		verifiedSigningCertificates = append(verifiedSigningCertificates, cert)
+	}
+	return verifiedSigningCertificates, nil
+}
+
+// verify verifies the trust in a top-down manner.
+//
+// References:
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verify(signerInfo *SignerInfo, opts *x509.VerifyOptions) (*x509.Certificate, error) {
+	// find signer certificate
+	cert := d.getCertificate(signerInfo.SignerIdentifier)
+	if cert == nil {
+		return nil, ErrCertificateNotFound
+	}
+
+	// verify signer certificate
+	certChains, err := cert.Verify(*opts)
+	if err != nil {
+		return cert, VerificationError{Detail: err}
+	}
+
+	// verify signature
+	if err := d.verifySignature(signerInfo, cert); err != nil {
+		return nil, err
+	}
+
+	// verify attribute
+	return cert, d.verifyAttributes(signerInfo, certChains)
+}
+
+// verifySignature verifies the signature with a trusted certificate.
+//
+// References:
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verifySignature(signerInfo *SignerInfo, cert *x509.Certificate) error {
+	// verify signature
+	algorithm := oid.ToSignatureAlgorithm(
+		signerInfo.DigestAlgorithm.Algorithm,
+		signerInfo.SignatureAlgorithm.Algorithm,
+	)
+	if algorithm == x509.UnknownSignatureAlgorithm {
+		return VerificationError{Message: "unknown signature algorithm"}
+	}
+
+	signed := d.Content
+	if len(signerInfo.SignedAttributes) > 0 {
+		encoded, err := asn1.MarshalWithParams(signerInfo.SignedAttributes, "set")
+		if err != nil {
+			return VerificationError{Message: "invalid signed attributes", Detail: err}
+		}
+		signed = encoded
+	}
+
+	if err := cert.CheckSignature(algorithm, signed, signerInfo.Signature); err != nil {
+		return VerificationError{Detail: err}
+	}
+	return nil
+}
+
+// verifyAttributes verifies the signed attributes.
+//
+// References:
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verifyAttributes(signerInfo *SignerInfo, chains [][]*x509.Certificate) error {
+	// verify attributes if present
+	if len(signerInfo.SignedAttributes) == 0 {
+		return nil
+	}
```

Comment:
As we discussed, we need to check the `ContentType of EncapsulatedContentInfo` here. If the type is not
```
id-data OBJECT IDENTIFIER ::= { iso(1) member-body(2)
       us(840) rsadsi(113549) pkcs(1) pkcs7(7) 1 }
```
then signerInfo.SignedAttributes MUST be present according to https://www.rfc-editor.org/rfc/rfc5652#section-5.3

Reply:
Updated.